src/pycell/impl_.rs

```diff
@@ -50,22 +51,43 @@ impl<M: PyClassMutability> PyClassMutability for ExtendsMutableAncestor<M> {
     type MutableChild = ExtendsMutableAncestor<MutableClass>;
 }
 
-#[derive(Debug, Copy, Clone, Eq, PartialEq)]
-struct BorrowFlag(usize);
+#[derive(Debug)]
+struct BorrowFlag(AtomicUsize);
 
 impl BorrowFlag {
-    pub(crate) const UNUSED: BorrowFlag = BorrowFlag(0);
-    const HAS_MUTABLE_BORROW: BorrowFlag = BorrowFlag(usize::MAX);
-    const fn increment(self) -> Self {
-        Self(self.0 + 1)
+    pub(crate) const UNUSED: usize = 0;
+    const HAS_MUTABLE_BORROW: usize = usize::MAX;
+    fn increment(&self) -> Result<(), PyBorrowError> {
+        let mut value = self.0.load(Ordering::Relaxed);
+        if value == BorrowFlag::HAS_MUTABLE_BORROW {
+            return Err(PyBorrowError { _private: () });
+        }
+        loop {
+            match self
+                .0
+                .compare_exchange(value, value + 1, Ordering::SeqCst, Ordering::SeqCst)
```

Comment:
Here I think that the `success`  needs to be `SeqCst` because if successful then a global ordering needs to be established to read any data protected from pyclass (similar to what I said to the mutable borrow case).

Given that, I think here the failure case might be ok to be `Acquire` purely because if it fails, the thread would then immediately retry and (hopefully) hit a `SeqCst` successful compare anyway. So maybe there's a case to use `Acquire` for the failure case, just to reduce the number of `SeqCst` ordering points. I have no idea whether that would actually produce a measurable improvement, so maybe best left as a FIXME subject to an attempt to measure contended borrows of a pyclass from multiple threads?

But of course, if the user uses `#[pyclass(frozen)]` then we will have none of this stuff so perhaps again there's an argument to just leave it all `SeqCst` and tell them to get performance by opting out of PyO3's borrow checking and using their own design. 😂

Reply:
See the latest commit, which I think is doing all the memory ordering correctly. I read through Mara Bos' chapter on this topic and I think this has the correct semantics. I also left comments to try to explain.

Reply:
Ok, having also read https://marabos.nl/atomics/memory-ordering.html#release-and-acquire-ordering, I have some different opinions.